### PR TITLE
Gracefully handle loop labels missing leading `'` in different positions

### LINKED ISSUE #81192: Account for missing `'` in break tags
[Given](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=9e04c8bd2fc029c8f1f219b224603807)

```rust
fn main() {
    'a: for _ in 0..1 {
        break 'a;
    }
    'b: for _ in 0..1 {
        break b;
    }
    c: for _ in 0..1 {
        break 'c;
    }
    d: for _ in 0..1 {
        break d;
    }
}
```

the current output is

```
error: expected identifier, found keyword `for`
 --> src/main.rs:8:8
  |
8 |     c: for _ in 0..1 {
  |        ^^^ expected identifier, found keyword

error: expected `<`, found reserved identifier `_`
 --> src/main.rs:8:12
  |
8 |     c: for _ in 0..1 {
  |      -     ^ expected `<`
  |      |
  |      tried to parse a type due to this type ascription
  |
  = note: `#![feature(type_ascription)]` lets you annotate an expression with a type: `<expr>: <type>`
  = note: see issue #23416 <https://github.com/rust-lang/rust/issues/23416> for more information

error[E0425]: cannot find value `b` in this scope
 --> src/main.rs:6:15
  |
6 |         break b;
  |               ^
  |               |
  |               not found in this scope
  |               help: a label with a similar name exists: `'b`

warning: unused label
 --> src/main.rs:5:5
  |
5 |     'b: for _ in 0..1 {
  |     ^^
  |
  = note: `#[warn(unused_labels)]` on by default

error[E0571]: `break` with value from a `for` loop
 --> src/main.rs:6:9
  |
6 |         break b;
  |         ^^^^^^^ can only break with a value inside `loop` or breakable block
  |
help: instead, use `break` on its own without a value inside this `for` loop
  |
6 |         break;
  |         ^^^^^

error: aborting due to 4 previous errors; 1 warning emitted
```

It should instead be 5 errors with appropriate suggestions to add the missing `'` to the tags.

### PR DESCRIPTION
**Pull request by estebank**
Fix #81192.

* Account for labels when suggesting `loop` instead of `while true`
* Suggest `'a` when given `a` only when appropriate
* Add loop head span to hir
* Tweak error for invalid `break expr`
* Add more misspelled label tests
* Avoid emitting redundant "unused label" lint
* Parse loop labels missing a leading `'`

Each commit can be reviewed in isolation.

### CODE REVIEW

File: src/test/ui/loops/loop-break-value.stderr [hunk 1]
```diff
@@ -18,101 +20,127 @@ LL |     'while_loop: while true {
...
-LL |         break LOOP;
-   |         ^^^^^^^^^^ can only break with a value inside `loop` or breakable block
+help: alternatively, you might have meant to use the available loop label
    |
-help: instead, use `break` on its own without a value inside this `for` loop
-   |
-LL |         break;
-   |         ^^^^^
+LL |             break 'for_loop 'for_loop;
+   |                             ^^^^^^^^^
```
Comment (estebank) [inline src/test/ui/loops/loop-break-value.stderr:143]: Mmm... I hadn't noticed this bad suggestion. It should be easy to fix.

File: compiler/rustc_resolve/src/late.rs [hunk 2]
```diff
@@ -2259,6 +2259,12 @@ impl<'a: 'ast, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
                 visit::walk_expr(self, expr);
             }
 
+            ExprKind::Break(None, Some(ref e)) => {
+                // We use this instead of `visit::walk_expr` to keep the parent expr around for
+                // better
```
Comment (oli-obk) [inline compiler/rustc_resolve/src/late.rs:2264]: this comment is cut off

File: compiler/rustc_hir/src/hir.rs [hunk 3]
```diff
@@ -1617,7 +1617,9 @@ pub enum ExprKind<'hir> {
     /// A conditionless loop (can be exited with `break`, `continue`, or `return`).
     ///
     /// I.e., `'label: loop { <block> }`.
-    Loop(&'hir Block<'hir>, Option<Label>, LoopSource),
+    ///
+    /// The `Span` is the loop header (`for x in y`/`while let pat = expr`).
+    Loop(&'hir Block<'hir>, Option<Label>, LoopSource, Span),
```
Comment (oli-obk) [inline compiler/rustc_hir/src/hir.rs:1622]: Some day we need to go through all of these tuple variants and make them struct variants... but not today :laughing:

File: src/test/ui/label/label_misspelled.rs [hunk 4]
```diff
@@ -15,6 +11,10 @@ fn main() {
         for_loop;
         //~^ ERROR cannot find value `for_loop` in this scope
     };
+    'LOOP: loop {
```
Comment (oli-obk) [inline src/test/ui/label/label_misspelled.rs:19]: why was this moved?

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: src/test/ui/label/label_misspelled.rs (hunk 4 above)
Comment (estebank) [reply src/test/ui/label/label_misspelled.rs:19]: To avoid the "unreachable code" warning (that doesn't trigger in ui tests unless you enable, but do trigger when testing manually).
